### PR TITLE
Add support for OAuth tokens

### PR DESCRIPTION
**Pull request by Wauplin**
OAuth with the Hub now support scopes as well. This means we can retrieve a token to manage repos, read files, write files, etc. on behalf of the user. More information here: https://huggingface.co/docs/hub/spaces-oauth#scopes

This PR add support for such tokens in the Gradio integration by adding a `gr.OAuthToken` object that can be injected in any function (similar to `gr.OAuthProfile`). I also decided to switch `gr.OAuthProfile` and `gr.OAuthToken` to dataclasses which are much more convenient for the user. I did it in a backward compatible way to not break the existing `gr.OAuthProfile` (that is currently a dict, but honestly that was a mistake from mine). I've also update the related documentation.

[Space Demo here:](https://huggingface.co/spaces/Wauplin/gradio-oauth-demo)

```py
import gradio as gr
from huggingface_hub import whoami

def hello(profile: gr.OAuthProfile | None) -> str:
    if profile is None:
        return "I don't know you."
    return f"Hello {profile.name}"

def list_organizations(oauth_token: gr.OAuthToken | None) -> str:
    if oauth_token is None:
        return "Please log in to list organizations."
    org_names = [org["name"] for org in whoami(oauth_token.token)["orgs"]]
    return f"You belong to {', '.join(org_names)}."

with gr.Blocks() as demo:
    gr.LoginButton()
    gr.LogoutButton()
    gr.Markdown().attach_load_event(hello, None)
    gr.Markdown().attach_load_event(list_organizations, None)
```

cc @abidlabs @coyotte508

### CODE REVIEW

File: guides/01_getting-started/03_sharing-your-app.md [hunk 1]
```diff
@@ -191,13 +191,15 @@ This is not the case by default for Safari, Chrome Incognito Mode.
 ### OAuth (Login via Hugging Face)
 
 Gradio supports OAuth login via Hugging Face. This feature is currently **experimental** and only available on Spaces.
-If allows to add a _"Sign in with Hugging Face"_ button to your demo. Check out [this Space](https://huggingface.co/spaces/Wauplin/gradio-oauth-demo)
-for a live demo.
+If allows to add a _"Sign in with Hugging Face"_ button to your demo. Check out [this Space](https://huggingface.co/spaces/Wauplin/gradio-oauth-demo) for a live demo.
```
Comment (abidlabs) [inline guides/01_getting-started/03_sharing-your-app.md:194]:
```suggestion
It allows you to add a _"Sign in with Hugging Face"_ button to your demo. Check out [this Space](https://huggingface.co/spaces/Wauplin/gradio-oauth-demo) for a live demo.
```

**Comment by abidlabs** [comment]
Awesome PR @Wauplin! Just a few nits:

1) For the example demo, I wonder if we can get away with requesting fewer permissions. The "manage all of your personal repos" scope is a bit intimidating.

image

2) It'd be good if `gr.OAuthToken` failed gracefully in non-Spaces environments, just like `gr.OAuthProfile`

Right now, we get this error, which crashes the Gradio app:

> requests.exceptions.HTTPError: 500 Server Error: Internal Server Error for url: https://huggingface.co/api/whoami-v2

**Comment by Wauplin** [comment]
> For the example demo, I wonder if we can get away with requesting fewer permissions. The "manage all of your personal repos" scope is a bit intimidating.

Ahah yes, good point. I wanted to showcase everything that was possible :smile: I updated to request only public profile and email.

image

> It'd be good if `gr.OAuthToken` failed gracefully in non-Spaces environments, just like gr.OAuthProfile

I will check what I can do :+1: Not so easy to have a fake token but that would still work locally.

**Comment by Wauplin** [comment]
Hey @abidlabs, I reworked a bit the oauth integration when working locally (see https://github.com/gradio-app/gradio/pull/6616/commits/8969ef99e1c23e4a902a5ec24d7713f78173c058). I think the easiest way to go is to use the token+profile of the user as mock data. If the user is not logged in (i.e. not `HF_TOKEN` or not `huggingface-cli login`) we raise an issue with some instructions on how to log in. This way, all subsequent methods will work as it would with any token. Please let me know what you think.

**Comment by abidlabs** [comment]
> I've having a lot of errors in the CI but I get this is because of the token issue in the tests, right?

Yes don't worry about those for now, we're going to fixing our CI as part of https://github.com/gradio-app/gradio/issues/6656

Taking a look at the rest of the PR now!

**Comment by Wauplin** [comment]
> Is the sample Space working for you @Wauplin? When I click the sign in with HF button and authorize the permissions, I get this:

Arf, thanks for letting me know @abidlabs! Looking at the logs it looks like an old session cookie was used which lead to this HTTP 500. I pushed a new commit (https://github.com/gradio-app/gradio/pull/6616/commits/b9a2224b684f13c25da58cdf28da4b10d061ac4a) to update the cookie secret key meaning that any old cookie will be invalidated. Users will have to re-login but that's not really a problem (+it's only for Spaces that will update their gradio version which is very rare).
Can you retry and let me know if you still have the same problem?

**Comment by abidlabs** [comment]
Yes, can confirm that the Space is working, thanks @Wauplin! Taking a look at the other pieces

**Comment by Wauplin** [comment]
Nice! Thanks for the review @abidlabs!

**Pull request merged**